### PR TITLE
Fixing the argument type definition of Marshal.GetNativeVariantForObject method in EnumVariantObject

### LINKED ISSUE #4120: [Accessibility] Missing the "linkLabel1 link" level for LinkLabel control in the inspect tool
* More info
1. The issue also can **reproduce on .NET 5.0 RC2.**
2. The issue can't reproduce on .NET SDK 3.1 and .NET 4.8.

* .NET Core Version: 
.NET SDK 6.0.100-alpha.1.20514.15

* Have you experienced this same bug with .NET Framework?: 
No

**Problem description:**
Missing the 'linkLabel1' link level for LinkLabel control in the inspect tool with UI Automation mode.

**Minimal repro:**
1.  Create a .NET Core project with a LinkLabel control.
2. Build and run the core project.
3. Observe the linkLabel1 control in the Inspect tool on UI Automation mode.

Or
1. Open the attached project.
2. Build and run the core project.
3. Observe the linkLabel1 control in the Inspect tool on UI Automation mode.

**Expected behavior:**
Have the correct levels displayed for linkLabel1 control in the Inspect tool.
ExpectResults

**Actual Behavior:**
Missing the "linkLabel1 link" level for LinkLabel control in the inspect tool.
NewIssueActual

### LINKED ISSUE #4121: [Accessibility] Missing the items displayed for CheckedListBox control in the inspect tool
* More info

1. The issue also can **reproduce on .NET 5.0 RC2.**
2. The issue can't reproduce on .NET SDK 3.1 and .NET 4.8.

* .NET Core Version: 
.NET SDK 6.0.100-alpha.1.20514.15

* Have you experienced this same bug with .NET Framework?: 
No

**Problem description:**
Missing the items displayed for CheckedListBox control in the inspect tool. 

**Minimal repro:**

1. Create a .NET Core project with a CheckedListBox control.
2. Build and run the core project.
3. Observe the checkedListBox1 control in the Inspect tool on UI Automation mode.

**Expected behavior:**
The items displayed correctly for checkedListBox1 control in the Inspect tool.
CheckedListBoxExpected

**Actual Behavior:**
Missing the items displayed for checkedListBox1 control in the inspect tool.
CheckedListBoxActual

### LINKED ISSUE #4130: [Accessibility]Inspect: The MSAA structure of NumericUpdown/DomainUpDown in Core is missing some children nodes
* .NET Core Version: 
.Net 6.0.100-alpha.1.20516.5

* Have you experienced this same bug with .NET Framework?: 
No

**More info:**
This issue cannot repro on .Net Core 3.1. 

**Minimal repro:**

1. Create a .NET Core project with a Label control and change the text to ‘numericUpDown’.
2. Add a NumericUpDown control to form.
3. Build and run the core project.
4. Open the Inspect tool and observe the NumericUpDown control on MSAA mode.

Or 

1. Extract the “MSAA_Numeric&Domain.zip” and open it in VS.
[MSAA_Numeric&Domain.zip](https://github.com/dotnet/winforms/files/5398625/MSAA_Numeric.Domain.zip)

2. Build and run the core project.
3. Open the Inspect tool and observe the NumericUpDown control on MSAA mode.

**Actual behavior:**
The MSAA structure of NumericUpdown/DomainUpDown in Core is missing some children nodes.
Core_5 0

**Expected behavior:**
The MSAA structure of NumericUpdown/DomainUpDown in Core is correct, it should be like Core 3.1:
core_3 1

### LINKED ISSUE #4172: Accessibility: In Inspect: MSAA, the child items of some controls cannot be displayed, and Children property gets error
* .NET Core Version: 
.Net 6.0.100-alpha.1.20516.5

* Have you experienced this same bug with .NET Framework?: 
No

* More Info
1. This issue cannot repro on .Net Core 3.1.
2. Affected controls: CheckedListBox, ListBox, ToolStrip, StatusStrip, ContextMenuStrip, BindingNavigator, DataGridView, ListView

**Minimal repro:**
1. Create a Winforms .Net Core app with controls: CheckedListBox/ListBox/ToolStrip/StatusStrip, ContextMenuStrip/BindingNavigator/DataGridView/ListView
2. Added items for these controls
3. Build and run app
4. Open the Inspect tool and observe these controls hierarchy on MSAA mode

**Actual behavior:**
Controls' items cannot be displayed in Inspect: MSAA 
image

**Expected behavior:**
Controls' items can be displayed in Inspect: MSAA hierarchy 
image

### PR DESCRIPTION
**Pull request by vladimir-krestov**
Fixes #4120 (`LinkLabel` accessiblity tree)
Fixes #4121 (`CheckedListBox` accessiblity tree)
Fixes #4130 (MSAA `UpDown` controls)
Fixes #4172 (MSAA `CheckedListBox`)

Regression from 0c4d3cb9198d9246e85af732c8fa81c3f3e0d88c

## Proposed changes

- Cast `currentChild` from `uint` to `int` before using in Marshal.GetNativeVariantForObject

## Customer Impact

- A user can see a correct accessibility tree for controls that don't support the UIA provider using Inspect tool

## Regression? 

- Yes (from 0c4d3cb9198d9246e85af732c8fa81c3f3e0d88c)

## Risk

- Minimal

### Before
- `CheckedListBox` doesn't have internal items AccessibleObjects in the accessibility tree
image

### After
- `CheckedListBox` has the correct accessibility tree with the internal items
image

## Test methodology 

- Manual testing
- CTI

## Accessibility testing  
- Using Inspect tool

 

## Test environment(s) 

- .Net Version: 6.0.0-alpha.1.20513.9
- Microsoft Windows [Version 10.0.19042.572]

###### Microsoft Reviewers: [Open in CodeFlow](http://wpcp.azurewebsites.net/CodeFlowProtocolProxy2.php?pullrequest=https://github.com/dotnet/winforms/pull/4168)

### CODE REVIEW

**Review by weltkante — COMMENTED** [review]
Did you identify what causes the erorr? You are changing the signature of a ComImport IUnknown interface, and of private implementation methods, neither of which are relevant (or visible) to external users

- ComImport means the definition is external (you don't generate the definition)
- IUnknown means you are implementing the definition without "reflection" support

I don't see any point in this changeset which could have an effect to external users of the interface. In particular since the official headers also use unsigned numbers there should never be any negative numbers appear anywhere.

If this fixes the problem I'm obviously missing something, and I'd like to know what :-)

**Comment by weltkante** [comment]
I think I see it now, you are not actually touching the line which was broken, which is what confuses me. You are changing a lot more than necessary to fix the bug.

The bug comes from `NextFromChildCollection` calling `Marshal.GetNativeVariantForObject` and being passed an `uint` instead of an `int` - by changing all the types to `int` (even those that shouldn't be according to native spec) you silently fixed the bug without touching the line causing the issue.

The "minmal" (and IMHO correct) fix would be just adding a typecast to `Marshal.GetNativeVariantForObject` argument instead of changing unrelated signatures from `uint` to `int`. The `IEnumVariant` method signatures don't need to be changed. Perhaps the `currentChild` should remain changed to `int`, can't evaluate that since I don't know too much context of its meaning.

TLDR: while the PR fixes the bug I'd argue the signature changes to `IEnumVariant` are wrong and should be reverted.

**Comment by RussKie** [comment]
I had a brief chat with @JeremyKuhne, and whilst our definitions match those in oaidl.h (`uint` == `ULONG`), there's _probably_ a contraction/expansion of the underlying value somewhere in the flow that leads to the loss/incorrect signedness...


> The "minmal" (and IMHO correct) fix would be just adding a typecast to `Marshal.GetNativeVariantForObject` argument

👍

**Comment by vladimir-krestov** [comment]
@weltkante, you are cool as always! You are right. 
The typecasting inside `Marshal.GetNativeVariantForObject` method is a sufficient change for the accessiblity tree to work correctly.

I really agree that changing `IEnumVariant` methods signature looks confusing because `IEnumVARIANT.Next` method has `ULONG` arguments ([A ULONG is a 32-bit unsigned integer (range: 0 through 4294967295 decimal)](https://docs.microsoft.com/openspecs/windows_protocols/ms-dtyp/32862b84-f6e6-40f9-85ca-c4faf985b822)) and it means (`uint == ULONG`). So it seems using **`uint`** for `IEnumVARIANT` is more correct.

Maybe @hughbe could say should we change **`currentChild`** type from **`uint`** to **`int`** in `EnumVariantObject`? Or can we keep it as is now?

**Comment by weltkante** [comment]
> there's _probably_ a contraction/expansion of the underlying value somewhere in the flow that leads to the loss/incorrect signedness

I didn't see anything of that sort, but `Marshal.GetNativeVariantForObject` will chose the type of the variant based on the type of the argument you pass in, so if the native code wants an "enumerator of signed integers" you need to pass a signed integer to get a `I4` variant.

The signedness of the `IEnumVARIANT` method arguments has nothing to do with the signedness of the values you are enumerating. Thats why it _may_ be correct to use `int` for `currentChild` (it represents the enumerated values?), but if you are confident that it never becomes negative its just as fine to keep it unsigned and only cast when handing it out through the native API.

**Review by weltkante — APPROVED** [review]

File: src/System.Windows.Forms/src/System/Windows/Forms/AccessibleObject.EnumVariantObject.cs [hunk 1]
```diff
@@ -179,7 +179,7 @@ private unsafe void NextFromChildCollection(uint celt, IntPtr rgVar, uint* pCelt
                 for (i = 0; i < celt && currentChild < childCount; ++i)
                 {
                     ++currentChild;
-                    Marshal.GetNativeVariantForObject(((object)currentChild), GetAddressOfVariantAtIndex(rgVar, i));
+                    Marshal.GetNativeVariantForObject(((object)(int)currentChild), GetAddressOfVariantAtIndex(rgVar, i));
```
Comment (RussKie) [inline src/System.Windows.Forms/src/System/Windows/Forms/AccessibleObject.EnumVariantObject.cs:182]: It would be good to add a comment why we cast here.

**Review by RussKie — APPROVED** [review]

**Review by weltkante — DISMISSED** [review]

**Comment by vladimir-krestov** [comment]
Testers approved the fix ✔️

**Review by RussKie — APPROVED** [review]
